core/local/atom_watcher.js

```diff
@@ -0,0 +1,81 @@
+/* @flow */
+
+const fse = require('fs-extra')
+const Promise = require('bluebird')
+
+const checksumer = require('./checksumer')
+const logger = require('../logger')
+const LinuxSource = require('./layers/linux')
+const Identifier = require('./layers/identifier')
+const ChecksumLayer = require('./layers/checksum')
+const Dispatcher = require('./layers/dispatcher')
+
+/*::
+import type Pouch from '../pouch'
+import type Prep from '../prep'
+import type EventEmitter from 'events'
+import type { Checksumer } from './checksumer'
+*/
+
+const log = logger({
+  component: 'AtomWatcher'
+})
+
+module.exports = class AtomWatcher {
+  /*::
+  syncPath: string
+  events: EventEmitter
+  checksumer: Checksumer
+  ensureDirInterval: *
+  running: Promise<void>
+  _runningResolve: ?Function
+  _runningReject: ?Function
+  source: LinuxSource
+  start: () => Promise<*>
+  stop: (force: ?bool) => Promise<*>
+  ensureDirSync: () => void
+  */
+
+  constructor (syncPath /*: string */, prep /*: Prep */, pouch /*: Pouch */, events /*: EventEmitter */) {
+    this.syncPath = syncPath
+    this.events = events
+    this.checksumer = checksumer.init()
+
+    // TODO detect platform to build the correct chain of layers
+    // TODO do we need a debounce layer (a port of awaitWriteFinish of chokidar)?
+    const dispatcher = new Dispatcher(prep, pouch, events)
+    const checksum = new ChecksumLayer(dispatcher, this.checksumer)
+    const identifier = new Identifier(checksum)
+    this.source = new LinuxSource(syncPath, identifier)
```

Comment:
I wonder about the *layer* abstraction...

- None of them actually accept/return exactly the same thing.
- Identifier and ChecksumLayer look like small single functions, while Linux and Dispatcher are huge.
- Do we actually need a full layer to assign ids while none of the next layers actually use it? Even so, shouldn't we just assign ids in `metadata.build*()`?
- Having to build the next Layer-compatible object to test any of them looks a bit complicated.
- All of them are mixing data, computation and side-effects, which doesn't make them easier to understand/test.

Reply:
I'm not really pleased with the layer abstraction. But I have a hard time to find better, and I still prefer this to the way the old watcher is organized. At least, each layer can be understood without to read other layers (even if it is a mess from a static types point of view).

> while Linux and Dispatcher are huge

They aren't huge, they are a lot smaller that the existing local analysis for moves.

Reply:
My apologies, *huge* was excessive.

Reply:
Regarding the `Identifier` layer, don't docs already have an id since they're built using `buildFile` / `buildDir` in the `LinuxSource` layer?

Reply:
Yes, it's the case now. It wasn't true when I wrote this code for `buildDir`, but it was fixed since then to match the behavior of `buildFile`.